Update requires grad property

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43743 Update aliasing in tensorexpr fuser
* #43742 use types in the IR instead of vmap
* #43636 Add inline call to profiling executor pipeline
* #43635 Use prim::TensorExprGroup interned symbol
* **#43634 Update requires grad property**

Because differentiable graphs detach the gradients of input Tensors, creating and inlining differentiable graphs changes the requires_grad property of tensors in the graph. In the legacy executor, this was not a problem as the Fuser would simply ignore the gradient property because it would be invariant that the LegacyExecutor only passed tensors with grad = False. This is not the case with the profiler, as the Fuser does it's own guarding. 

Updating the type also helps with other typechecks, e.g. the ones specializing the backward, and with debugging the graph.

Other possibilities considered were:
- Fuser/Specialize AutogradZero always guards against requires_grad=False regardless of the profiled type
- Re-profile forward execution of differentiable graph

Differential Revision: [D23358803](https://our.internmc.facebook.com/intern/diff/D23358803)

Commit:
Update requires grad property

[ghstack-poisoned]

Commit:
Update on "Update requires grad property"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/runtime/autodiff.cpp:
```diff
@@ -841,6 +841,12 @@ Gradient differentiate(std::shared_ptr<Graph>& graph) {
   lambdaLiftReverse(grad_desc, rev_info);
   packReturnValuesIntoTuple(grad_desc.df);
+
+  // we have created a differentiable forward graph
+  // which will be run with tensors that have their gradients detached,
+  // so profiled types will have outdated requires_grad=True, update the
+  // requires_grad property
+  UpdateDifferentiableGraphRequiresGrad(grad_desc.f, false);
```
Is it true for inputs/outputs of the graph as well?

eellison replied:
it is true but the only downstream user of this property is the profiling executor, which only looks at prim::profile nodes. i can update the. comment

ZolotukhinM reviewed (approved):
Looks good!

A minor nit: the commit message header makes the change appear much more global and intrusive than it actually is. Maybe add a couple of words saying that we're just adding a jit pass or something like that?

Commit:
Update on "Update requires grad property"


Because differentiable graphs detach the gradients of input Tensors, creating and inlining differentiable graphs changes the requires_grad property of tensors in the graph. In the legacy executor, this was not a problem as the Fuser would simply ignore the gradient property because it would be invariant that the LegacyExecutor only passed tensors with grad = False. This is not the case with the profiler, as the Fuser does it's own guarding. 

Updating the type also helps with other typechecks, e.g. the ones specializing the backward, and with debugging the graph.

Other possibilities considered were:
- Fuser/Specialize AutogradZero always guards against requires_grad=False regardless of the profiled type
- Re-profile forward execution of differentiable graph

Differential Revision: [D23358803](https://our.internmc.facebook.com/intern/diff/D23358803)

[ghstack-poisoned]

Commit:
Update on "Update requires grad property"


Because differentiable graphs detach the gradients of input Tensors, creating and inlining differentiable graphs changes the requires_grad property of tensors in the graph. In the legacy executor, this was not a problem as the Fuser would simply ignore the gradient property because it would be invariant that the LegacyExecutor only passed tensors with grad = False. This is not the case with the profiler, as the Fuser does it's own guarding. 

Updating the type also helps with other typechecks, e.g. the ones specializing the backward, and with debugging the graph.

Other possibilities considered were:
- Fuser/Specialize AutogradZero always guards against requires_grad=False regardless of the profiled type
- Re-profile forward execution of differentiable graph

Differential Revision: [D23358803](https://our.internmc.facebook.com/intern/diff/D23358803)

[ghstack-poisoned]

Commit:
Update on "Update requires grad property"


Because differentiable graphs detach the gradients of input Tensors, creating and inlining differentiable graphs changes the requires_grad property of tensors in the graph. In the legacy executor, this was not a problem as the Fuser would simply ignore the gradient property because it would be invariant that the LegacyExecutor only passed tensors with grad = False. This is not the case with the profiler, as the Fuser does it's own guarding. 

Updating the type also helps with other typechecks, e.g. the ones specializing the backward, and with debugging the graph.

Other possibilities considered were:
- Fuser/Specialize AutogradZero always guards against requires_grad=False regardless of the profiled type
- Re-profile forward execution of differentiable graph

Differential Revision: [D23358803](https://our.internmc.facebook.com/intern/diff/D23358803)

[ghstack-poisoned]

Commit:
Update on "Update requires grad property"


Because differentiable graphs detach the gradients of input Tensors, creating and inlining differentiable graphs changes the requires_grad property of tensors in the graph. In the legacy executor, this was not a problem as the Fuser would simply ignore the gradient property because it would be invariant that the LegacyExecutor only passed tensors with grad = False. This is not the case with the profiler, as the Fuser does it's own guarding. 

Updating the type also helps with other typechecks, e.g. the ones specializing the backward, and with debugging the graph.

Other possibilities considered were:
- Fuser/Specialize AutogradZero always guards against requires_grad=False regardless of the profiled type
- Re-profile forward execution of differentiable graph

Differential Revision: [D23358803](https://our.internmc.facebook.com/intern/diff/D23358803)

[ghstack-poisoned]

Commit:
Update on "Update requires grad property"


Because differentiable graphs detach the gradients of input Tensors, creating and inlining differentiable graphs changes the requires_grad property of tensors in the graph. In the legacy executor, this was not a problem as the Fuser would simply ignore the gradient property because it would be invariant that the LegacyExecutor only passed tensors with grad = False. This is not the case with the profiler, as the Fuser does it's own guarding. 

Updating the type also helps with other typechecks, e.g. the ones specializing the backward, and with debugging the graph.

Other possibilities considered were:
- Fuser/Specialize AutogradZero always guards against requires_grad=False regardless of the profiled type
- Re-profile forward execution of differentiable graph

Differential Revision: [D23358803](https://our.internmc.facebook.com/intern/diff/D23358803)

[ghstack-poisoned]
